### PR TITLE
Fix: Handle client connection reset for keep-alive connections

### PR DESCRIPTION
**Pull request by tisba**
See https://github.com/ruby/webrick/pull/86

`TCPSocket#eof?` will raise `Errno::ECONNRESET` if the connection has been reset. This is currently not handled properly in `WEBrick::HTTPServer#run`. This was reported by @carlhoerberg, but incorrectly addressed in https://github.com/ruby/webrick/pull/86.

Preconditions to hit this issue:

* Client has to request with `Connection: keep-alive`
* WEBrick has to respect that (e.g. errors / 404s do close the connection)
* Client resets the connection after the first request

This will leave the server believe that the connection is still good and fail in `WEBrick::HTTPServer#run` when testing `TCPSocket#eof?` in the next iteration of `#run`.

Testing this is a little tricky, here is how I approached it:

* start a test server, using `WEBrick::HTTPServer`
* setup a success response on `/`
* do a single request with `Connection: keep-alive`
* after reading the response body, fish the `TCPSocket` object out of the `Net::HTTP` instance and configure it to linger on close and forcefully reset the connection, dropping all remaining data. This will send a `RST` to the server
* to avoid `TestWEBrick` shutting down, before we hit the issue, we need to wait a little bit

### CODE REVIEW

**Comment by tisba** [comment]
Can somebody kick of the test workflow? @olleolleolle maybe? :)

**Review by jeremyevans — APPROVED** [review]
Thank you for working on this!  The approach you used for fixing the issue seems best to me.

**Comment by tisba** [comment]
Got the error again yesterday (using webrick a lot in little local projects) and got reminded of this: Is there anything left to be done here for me, @jeremyevans?

**Comment by jeremyevans** [comment]
@tisba Nothing left for you to do.  I've approved this, so another developer should just need to review and merge it.  I would recommend against using webrick, though.  Is there a reason you couldn't use puma?

**Comment by tisba** [comment]
Let me rephrase and clarify: Anything that I work on and is production-grade uses puma. But I have some very tiny toy projects where webrick is great, because no native extensions, minimal dependencies etc. I also still use an alias to `ruby -run -e httpd . -p 9090` almost daily – old habits just die hard :)

**Comment by tisba** [comment]
Hey @hsbt 👋 Could you maybe take a look at this? 🙇
